### DISCUSSION TITLE
MSP_SET_HEAD Message 211

### DISCUSSION THREAD

**Discussion by jongell** [iNavFlight/inav#9606] [Q&A] [answered]
Hi all is this message also available when I engage Cruise Mode in Fixed Wing?

Also setting the heading involves converting it to -180° - 180°, 2 bytes and signed?

Best, 

Jonas

**Comment by sensei-hacker** [comment]
Looks like it does apply to fixed wing and is a uint16, ranging from 0 to 36000

Test to confirm.

https://github.com/iNavFlight/inav/blob/af0b3f42c0e70d7d8418068204c44002fdb20f21/src/main/fc/fc_msp.c#L1734

**Reply by jongell** [reply]
Thanks a lot! Will test and come back. When I request the gps groundcourse it comes back as decimal degrees so 0-3600

Let me try both 36000 and 3600

**Comment by jongell** [comment]
It looks like the grouncourse is going to be uint16 and centidegrees: 0-36000 https://github.com/iNavFlight/inav/blob/a90203cac97efb9c6a9879002a1371c32ccc250a/src/main/msp/msp_protocol_v2_sensor_msg.h#L54

Unfortunately for fixed Wing MSP_SET_HEAD is not working. 

I tested today with some findings:

NAV HEADING HOLD is not available for fixed wing, at least when I try to activate the mode it flips back into acro. 

NAV CRUISE activates NAV COURSE HOLD and NAV ALTHOLD. 

So the solution would be to update the course when a MSP_SET_HEAD message comes in. But It looks like that the way a course adjustment works is the following: 

When the user inputs roll or yaw it temporarily goes out of course hold and resotores it when the user lets go: https://github.com/iNavFlight/inav/blob/af0b3f42c0e70d7d8418068204c44002fdb20f21/src/main/navigation/navigation.c#L1186

**Reply by jongell** [reply]
Would it be possible to use this: posControl.cruise.course = https://github.com/iNavFlight/inav/blob/af0b3f42c0e70d7d8418068204c44002fdb20f21/src/main/navigation/navigation.c#L1188 inside MSP_SET_HEAD ?

**Reply by jongell** [reply, marked as answer]
Turns out yes. That works. I adjusted the MSP_SET_HEAD switch case in fc_msp.c : 

case MSP_SET_HEAD:
        if (sbufReadU16Safe(&tmp_u16, src)){
            updateHeadingHoldTarget(tmp_u16);
            posControl.cruise.course = (int32_t)tmp_u16;
        }
        else
            return MSP_RESULT_ERROR;
        break;
        
Then the course is deg * 100, 2 bytes unsigned and as long as user is not yawing or rolling it immediately will take that as the target course during cruise mode. according to 

https://github.com/iNavFlight/inav/blob/a90203cac97efb9c6a9879002a1371c32ccc250a/src/main/navigation/navigation.c#L1176
